Platform view blur clipping - Rounded Rect (iOS)

Hari-07 opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

This PR aims to fix blurs on platform views not being clipped on ios. Currently the clips were not forwarded to the mutator stack and thus just ignored. The frame the blur is in was being used which would always behave like a clip rect ignoring any corner radii or clip paths etc

Now all rrect, rse and path clips are being forwarded to the mutator stack. 

**Currently only rounded rect has been implemented though, and even in that just a single uniform corner radii, which seems to be most of the use cases from users who have run into this issue.** 


Example: 
```dart
Widget build(BuildContext context) {
    return ClipRRect(
      borderRadius: BorderRadius.circular(50),
      child: Container(
        width: 200,
        height: 200,
        decoration: BoxDecoration(
          color: Colors.black.withOpacity(0.2),
          borderRadius: BorderRadius.circular(500),
        ),
        child: BackdropFilter(
          filter: ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
          child: const Center(
            child: Text('Blurred block 1'),
          ),
        ),
      ),
    );
```
Before:
<img src="https://github.com/user-attachments/assets/229c82a0-70de-4e23-841e-2cbea92cc79f" height="600">

After:
<img src="https://github.com/user-attachments/assets/537f4215-5f0f-4cd6-aa09-e058d133e8ac" height="600">


*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*
https://github.com/flutter/flutter/issues/115926

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

Hari-07 committed:
Push the backdrop clips into embedded view

Hari-07 committed:
Get the clip data into the mutator stack

Hari-07 committed:
Get rounded rect working

Hari-07 committed:
Update the corner radius correctly on hot reload

LukeMoody01 commented:
Tested this fix in our production app (heavy usage of blurred overlays over platform views). The clipping works perfectly, but we're seeing significant performance impact:

- **Debug builds**: 2-3ms per raster frame (~15-25% slower)
- **Release builds**: 1-2ms per raster frame (~8-15% slower)

The bottleneck seems to be:
1. Platform view layer creation for each blur instance on every frame
2. SkRRect → CALayer geometry conversion per frame
3. No caching when clip geometry hasn't changed

Potential optimisations:
- Cache clip layers between frames when geometry is static
- Batch multiple BackdropFilters into a single platform view
- GPU shader-based clipping instead of CALayer composition

Happy to provide profiling data or test optimisation approaches if there's interest.

Test setup: Flutter 3.35.7 baseline vs custom build with this PR, iPhone 16 Pro (iOS 18.6.2)

Hari-07 committed:
Remove print statements

Hari-07 commented:
@LukeMoody01 diving a bit more into the bottlenecks and optimisations you have mentioned - any chance this was AI generated? Because this isn't really making sense to me based off of the code. For example

> No caching when clip geometry hasn't changed.
> Platform view layer creation for each blur instance on every frame

As I understand `@property(nonatomic) NSMutableArray<UIVisualEffectView*>* backdropFilterSubviews;` is storing the filters and those are being reused 
```
if (self.backdropFilterSubviews.count <= index) {
      backdropFilterView = filter.backdropFilterView;
      [self addSubview:backdropFilterView];
      [self.backdropFilterSubviews addObject:backdropFilterView];
    } else {
      [filter updateVisualEffectView:self.backdropFilterSubviews[index]];
    }
```

I did have some stray logs I had added in this PR that I have now removed, can you try rerunning your performance benchmark? I think the perf hit just came from those rather than anything else

Hari-07 committed:
Implement mutator stack unit tests

Hari-07 committed:
ios test for cliprrect

Hari-07 committed:
Formatting unit tests

Hari-07 committed:
Fix missing cases in android

Hari-07 commented:
> Bottlenecks were indeed AI-generated based on your git diff, engine code isn’t my usual domain, so I’d take those suggestions with a healthy grain of salt.

Please avoid these IMO. While its obviously helpful to test it out and share the performance you saw, if unfamiliar its better not to comment on bottlenecks and fixes rather than use AI and give wrong suggestions which ended up wasting my time trying to make sense of it

gmackall reviewed (commented):
From android triage: just noting that the small android portion looks reasonable, to get it out of the android triage queue

hellohuanlin commented on engine/src/flutter/flow/embedded_views.h:
```diff
@@ -109,6 +171,11 @@ class Mutator {
       case MutatorType::kClipPath:
       case MutatorType::kClipRRect:
       case MutatorType::kClipRSE:
+      // TODO: Figure out if this matters
```
reminder on this before landing

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsController.mm:
```diff
@@ -576,8 +583,21 @@ - (void)applyMutators:(const flutter::MutatorsStack&)mutatorsStack
         UIVisualEffectView* visualEffectView = [[UIVisualEffectView alloc]
             initWithEffect:[UIBlurEffect effectWithStyle:UIBlurEffectStyleLight]];
+
+        // TODO: Technically we should be using the largest corner radius, but currently
```
this should be tracked by a github issue

Hari-07 replied:
Filed as https://github.com/flutter/flutter/issues/179126

hellohuanlin commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsController.mm:
```diff
@@ -586,6 +606,31 @@ - (void)applyMutators:(const flutter::MutatorsStack&)mutatorsStack
+        [pendingClipRRects addObject:clip];
+      }
+      case flutter::MutatorType::kBackdropClipRse: {
+        // TODO: Pending Implementation
```
same here

Hari-07 committed:
Rename all RSE to RSuperellipse

Hari-07 committed:
Run formatter

Hari-07 committed:
Rename PushClipToEmbedderStack to PushClipToEmbeddedNativeViewMutatorStack

Hari-07 committed:
Update IsClipType in embedded_views

Hari-07 committed:
File github issues for later todos

hellohuanlin reviewed (approved):
LGTM!

gmackall committed:
Merge branch 'master' into platform-view-blur-clipping

LouiseHsu commented:
Unfortunately I think there are legit failures, if you look at 
https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20Engine%20Drone/2067659/overview

```
  /Volumes/Work/s/w/ir/cache/builder/engine/src/out/ci/host_release_arm64_tests/flow_unittests ClipPathLayerTest.ChildOutsideBounds passed 0 / 2 times.
  /Volumes/Work/s/w/ir/cache/builder/engine/src/out/ci/host_release_arm64_tests/flow_unittests ClipPathLayerTest.FullyContainedChild passed 0 / 2 times.
  /Volumes/Work/s/w/ir/cache/builder/engine/src/out/ci/host_release_arm64_tests/flow_unittests ClipPathLayerTest.LayerCached passed 0 / 2 times.
  /Volumes/Work/s/w/ir/cache/builder/engine/src/out/ci/host_release_arm64_tests/flow_unittests ClipPathLayerTest.OpacityInheritance passed 0 / 2 times.
  /Volumes/Work/s/w/ir/cache/builder/engine/src/out/ci/host_release_arm64_tests/flow_unittests ClipPathLayerTest.OpacityInheritancePainting passed 0 / 2 times.
  /Volumes/Work/s/w/ir/cache/builder/engine/src/out/ci/host_release_arm64_tests/flow_unittests ClipPathLayerTest.OpacityInheritanceSaveLayerPainting passed 0 / 2 times.
  /Volumes/Work/s/w/ir/cache/builder/engine/src/out/ci/host_release_arm64_tests/flow_unittests ClipPathLayerTest.OvalPathReducesToClipRRect passed 0 / 2 times.
  ```
  
  Which definitely seems related to your changes. Can you take a look?

Hari-07 committed:
Fix null platform view case

Hari-07 committed:
Merge remote-tracking branch 'upstream/master' into platform-view-blur-clipping

Hari-07 committed:
Fix test

LouiseHsu reviewed (approved):
Looks good to me! Thanks for addressing the tests~

Hari-07 committed:
Merge branch 'master' of https://github.com/flutter/flutter into platform-view-blur-clipping

Hari-07 committed:
Merge branch 'master' of https://github.com/flutter/flutter into platform-view-blur-clipping

Hari-07 committed:
Merge branch 'master' of https://github.com/flutter/flutter into platform-view-blur-clipping
